### PR TITLE
filter: Add SimpleHttpCache

### PR DESCRIPTION
**Pull request by toddmgreer**
Description: Add CacheFilter's first rudimentary storage plugin
Risk Level: Low; this code is only used in unit tests
Testing: unit tests
Docs Changes: n/a
Release Notes: n/a
#868

### CODE REVIEW

**Comment by mattklein123** [comment]
Per https://github.com/envoyproxy/envoy/issues/9977, can we figure out how to wire the cache extensions into the extension build system so that the user can select which extensions they want to include in the build? cc @enozcan. Thank you!

/wait

**Comment by toddmgreer** [comment]
Used extension build system to make SimpleHttpCache's build disablable.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+LookupContextPtr SimpleHttpCache::makeLookupContext(LookupRequest&& request) {
+  return std::make_unique<SimpleLookupContext>(*this, std::move(request));
+}
+
+void SimpleHttpCache::updateHeaders(LookupContextPtr&& lookup_context,
+                                    Http::HeaderMapPtr&& response_headers) {
+  ASSERT(lookup_context);
+  ASSERT(response_headers);
+  // TODO(toddmgreer): Support updating headers.
+  ASSERT(false);
```
Comment (mattklein123) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Crash with NOT_IMPLEMENTED

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    body_.add(chunk);
+    if (end_stream) {
+      commit();
+    } else {
+      ready_for_next_chunk(true);
+    }
+  }
+
+  void insertTrailers(const Http::HeaderMap&) override {
+    ASSERT(false); // TODO(toddmgreer): support trailers
```
Comment (mattklein123) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: NOT_IMPLEMENTED

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+  void getBody(const AdjustedByteRange& range, LookupBodyCallback&& cb) override {
+    ASSERT(range.end() <= body_.length(), "Attempt to read past end of body.");
+    cb(std::make_unique<Buffer::OwnedImpl>(&body_[range.begin()], range.length()));
+  }
+
+  void getTrailers(LookupTrailersCallback&& cb) override {
+    // TODO(toddmgreer): Support trailers.
+    ASSERT(false, "We didn't say there were trailers.");
+    cb(nullptr);
```
Comment (mattklein123) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: NOT_IMPLEMENTED

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc (hunk 2 above)
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Done.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc (hunk 1 above)
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Done.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+InsertContextPtr SimpleHttpCache::makeInsertContext(LookupContextPtr&& lookup_context) {
+  ASSERT(lookup_context != nullptr);
+  return std::make_unique<SimpleInsertContext>(*lookup_context, *this);
+}
+
+CacheInfo SimpleHttpCache::cacheInfo() const {
+  CacheInfo cache_info;
+  cache_info.name_ = "envoy.extensions.http.cache.simple";
```
[the remark this answers is not in the document]
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Done.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+public:
+  SimpleHttpCacheFactory() : HttpCacheFactory("envoy.extensions.http.cache.simple") {}
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return std::make_unique<
+        envoy::source::extensions::filters::http::cache::SimpleHttpCacheConfig>();
+  }
+  HttpCache& getCache() override { return cache_; }
+
+private:
+  SimpleHttpCache cache_;
```
[the remark this answers is not in the document]
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: There are a variety of reasonable ways to partition a cache. The simplest way is to not partition it, and that makes sense for SimpleHttpCache.

I have now added back in the behavior of passing the config to getCache to support different partitionings. (I had removed it to make the HttpCache PR independently reviewable from the API PR.)

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc (hunk 3 above)
Comment (toddmgreer) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Done.

**Pull request merged**